File: OrcScala/src/orc/lib/state/Ref.java
//
// Ref.java -- Java class Ref
// Project OrcScala
//
// $Id$
//
// Copyright (c) 2009 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state;

import java.util.LinkedList;
import java.util.Queue;

import orc.Handle;
import orc.error.runtime.ArityMismatchException;
import orc.error.runtime.TokenException;
import orc.lib.state.types.RefType;
import orc.types.Type;
import orc.values.sites.TypedSite;
import orc.values.sites.compatibility.Args;
import orc.values.sites.compatibility.DotSite;
import orc.values.sites.compatibility.EvalSite;
import orc.values.sites.compatibility.SiteAdaptor;

/**
 * Rewritable mutable reference. The reference can be initialized with a value,
 * or left initially empty. Read operations block if the reference is empty.
 * Write operations always succeed.
 *
 * @author dkitchin
 */
public class Ref extends EvalSite implements TypedSite {

	@Override
	public Object evaluate(final Args args) throws TokenException {

		// If we were passed an argument, store it in the ref as an initial
		// value
		if (args.size() == 1) {
			return new RefInstance(args.getArg(0));
		}
		// With no args, create an initially empty reference
		else if (args.size() == 0) {
			return new RefInstance();
		} else {
			throw new ArityMismatchException(1, args.size());
		}
	}

	@Override
	public Type orcType() {
		return RefType.getBuilder();
	}

	public static class RefInstance extends DotSite {

		protected Queue<Handle> readQueue;
		Object contents;

		public RefInstance() {
			this.contents = null;

			/*
			 * Note that the readQueue also signals whether the reference has
			 * been assigned. If it is non-null (as it is initially), the
			 * reference is unassigned. If it is null, the reference has been
			 * assigned.
			 *
			 * This allows the reference to contain a null value if
			 * needed, and it also frees the memory associated with the read
			 * queue once the reference has been assigned.
			 */
			this.readQueue = new LinkedList<Handle>();
		}

		/*
		 * Create the reference with an initial value already assigned. In this
		 * case, we don't need a reader queue.
		 */
		public RefInstance(final Object initial) {
			this.contents = initial;
			this.readQueue = null;
		}

		@Override
		protected void addMembers() {
			addMember("read", new readMethod());
			addMember("write", new writeMethod());
			addMember("readD", new SiteAdaptor() {
				@Override
				public void callSite(final Args args, final Handle caller) throws TokenException {
					synchronized (RefInstance.this) {
						if (readQueue != null) {
							caller.halt();
						} else {
							caller.publish(object2value(contents));
						}
					}
				}
                @Override
                public boolean nonBlocking() { return true; }
			});
		}

		protected class readMethod extends SiteAdaptor {
			@Override
			public void callSite(final Args args, final Handle reader) {
				synchronized (RefInstance.this) {
					/*
					 * If the read queue is not null, the ref has not been set.
					 * Add this caller to the read queue.
					 */
					if (readQueue != null) {
						reader.setQuiescent();
						readQueue.add(reader);
					}
					/* Otherwise, return the contents of the ref */
					else {
						reader.publish(object2value(contents));
					}
				}
			}
		}

		protected class writeMethod extends SiteAdaptor {
			@Override
			public void callSite(final Args args, final Handle writer) throws TokenException {
				synchronized (RefInstance.this) {

					final Object val = args.getArg(0);

					/* Set the contents of the ref */
					contents = val;

					/*
					 * If the read queue is not null, the ref has not yet been
					 * set.
					 */
					if (readQueue != null) {

						/*
						 * Wake up all queued readers and report the written
						 * value to them.
						 */
						for (final Handle reader : readQueue) {
							reader.publish(object2value(val));
						}

						/*
						 * Null out the read queue. This indicates that the ref
						 * has been written. It also allowed the associated
						 * memory to be reclaimed.
						 */
						readQueue = null;
					}

					/* A write always succeeds and publishes a signal. */
					writer.publish(object2value(signal()));
				}
			}
            @Override
            public boolean nonBlocking() { return true; }
		}

		@Override
		public String toString() {
			return "Ref(" + contents + ")";
		}
	}

    @Override
    public boolean nonBlocking() { return true; }
}

File: OrcScala/src/orc/lib/state/Ref.scala
//
// Ref.java -- Java class Ref
// Project OrcScala
//
// Copyright (c) 2017 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state

import java.util.{LinkedList, Queue}

import scala.collection.JavaConverters.collectionAsScalaIterableConverter

import orc.{
  DirectInvoker,
  MaterializedCallContext,
  OrcRuntime,
  VirtualCallContext
}
import orc.error.runtime.ArityMismatchException
import orc.lib.state.types.RefType
import orc.run.distrib.{AbstractLocation, ClusterLocations, DOrcPlacementPolicy}
import orc.types.Type
import orc.values.{FastObject, Signal}
import orc.values.sites.{
  FunctionalSite,
  NonBlockingSite,
  PartialSite0Simple,
  Site0Simple,
  Site1Simple,
  SiteMetadata,
  TargetClassAndArgumentClassSpecializedInvoker,
  TotalSiteBase,
  TypedSite
}

/** Rewritable mutable reference. The reference can be initialized with a value,
  * or left initially empty. Read operations block if the reference is empty.
  * Write operations always succeed.
  *
  * @author dkitchin
  */
object Ref
    extends TotalSiteBase
    with TypedSite
    with SiteMetadata
    with FunctionalSite {
  override val inlinable = true
  def getInvoker(runtime: OrcRuntime, args: Array[AnyRef]): DirectInvoker = {
    args.size match {
      case 1 => invoker(this, args: _*)((_, args) => new Ref.Instance(args(0)))
      case 0 => invoker(this, args: _*)((_, args) => new Ref.Instance())
      case _ =>
        new TargetClassAndArgumentClassSpecializedInvoker(this, args)
        with DirectInvoker {
          @throws[Throwable]
          def invokeDirect(target: AnyRef, arguments: Array[AnyRef]): AnyRef = {
            throw new ArityMismatchException(1, args.size)
          }
        }
    }
  }

  override def orcType(): Type = RefType.getBuilder

  private val members = FastObject.members("read", "write", "readD")

  class Instance() extends FastObject(members) with DOrcPlacementPolicy {

    protected var readQueue: Queue[MaterializedCallContext] =
      new LinkedList[MaterializedCallContext]()

    var contents: AnyRef = null

    /*
     * Create the reference with an initial value already assigned. In this
     * case, we don't need a reader queue.
     */
    def this(initial: AnyRef) = {
      this()
      this.contents = initial
      this.readQueue = null
    }

    protected val values = Array(
      new readMethod(),
      new writeMethod(),
      new PartialSite0Simple with NonBlockingSite {
        override val inlinable = true
        // readD
        override def eval() = {
          Instance.this.synchronized {
            if (readQueue != null) {
              None
            } else {
              Some(contents)
            }
          }
        }
      }
    )

    protected class readMethod extends Site0Simple {
      override val inlinable = true

      override def eval(reader: VirtualCallContext) = {
        Instance.this.synchronized {
          /*
           * If the read queue is not null, the cell has not been set.
           * Add this caller to the read queue.
           */

          if (readQueue != null) {
            val mat = reader.materialize()
            mat.setQuiescent()
            readQueue.add(mat)
            reader.empty
          } else
            /* Otherwise, return the contents of the cell */ {
              reader.publish(contents)
            }
        }
      }

    }

    protected class writeMethod extends Site1Simple[AnyRef]() {
      override val inlinable = true

      override def eval(writer: VirtualCallContext, v: AnyRef) = {
        Instance.this.synchronized {
          /* Set the contents of the cell */

          contents = v
          /* A write publishes a signal. */
          val r1 = writer.publish(Signal)
          if (readQueue != null) {
            val rq: Queue[MaterializedCallContext] = readQueue
            /*
             * Null out the read queue. This indicates that the cell
             * has been written. It also allowed the associated
             * memory to be reclaimed.
             */

            readQueue = null
            /*
             * Wake up all queued readers and report the written
             * value to them.
             */

            rq.asScala.foldLeft(r1)(_.publish(_, v))
          } else {
            r1
          }
        }
      }
    }

    override def toString(): String = "Ref(" + contents + ")"

    override def permittedLocations[L <: AbstractLocation](
        locations: ClusterLocations[L]): scala.collection.immutable.Set[L] =
      locations.hereSet
  }
}
